File: test/test_cli_progress.py
#!/usr/bin/env python3
"""Python CLI progress is opt-in off-TTY and never contaminates artifacts."""

from __future__ import annotations

import json
from pathlib import Path
import subprocess
import sys
import tempfile


class _TerminalCapture:
    def __init__(self) -> None:
        self.value = ""

    def isatty(self) -> bool:
        return True

    def write(self, value: str) -> int:
        self.value += value
        return len(value)

    def flush(self) -> None:
        return None


def verify_adaptive_terminal() -> None:
    """Exercise the delayed, single-line TTY projection without sleeping."""

    import archbird.cli as cli

    capture = _TerminalCapture()
    ticks = iter((0.0, 0.1, 1.1, 1.3, 1.4))
    original_stderr = cli.sys.stderr
    original_monotonic = cli.time.monotonic
    try:
        cli.sys.stderr = capture
        cli.time.monotonic = lambda: next(ticks)
        progress = cli._Progress("auto")
        progress.emit({"phase": "providers", "provider": "empty", "state": "start", "total": 0})
        progress.emit(
            {
                "phase": "providers",
                "provider": "syntax:tree-sitter:javascript",
                "state": "progress",
                "completed": 5,
                "total": 10,
            }
        )
        progress.emit({"phase": "joining", "state": "start"})
        progress.finish()
    finally:
        cli.time.monotonic = original_monotonic
        cli.sys.stderr = original_stderr
    if "empty" in capture.value or "0/0" in capture.value:
        raise AssertionError(capture.value)
    if "5/10 files (50%)" not in capture.value:
        raise AssertionError(capture.value)
    if not capture.value.startswith("\r") or capture.value.count("\n") != 1:
        raise AssertionError(f"TTY progress was not one updating line: {capture.value!r}")


def run(
    root: Path,
    mode: str,
    *,
    cache_dir: Path | None = None,
    cache_max_bytes: int | None = None,
    check: bool = True,
) -> subprocess.CompletedProcess[bytes]:
    fixture = root / "test/fixtures/map_base"
    return subprocess.run(
        [
            sys.executable,
            "-m",
            "archbird",
            "map",
            str(fixture),
            "--config",
            str(fixture / "archbird.json"),
            "--progress",
            mode,
            "--python-provider-timeout",
            "30",
            *(
                ["--no-cache"]
                if cache_dir is None
                else ["--cache-dir", str(cache_dir)]
            ),
            *(
                []
                if cache_max_bytes is None
                else ["--cache-max-bytes", str(cache_max_bytes)]
            ),
            "--format",
            "json",
            *(["--check"] if check else []),
        ],
        check=True,
        cwd=root,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )


def main() -> int:
    if len(sys.argv) != 2:
        raise SystemExit("usage: test_cli_progress.py REPOSITORY")
    root = Path(sys.argv[1]).resolve()
    always = run(root, "always")
    automatic = run(root, "auto")
    streamed = run(root, "auto", check=False)
    if always.stdout != automatic.stdout:
        raise AssertionError("progress mode changed canonical Map bytes")
    if streamed.stdout != always.stdout:
        raise AssertionError("streaming changed canonical Map bytes")
    document = json.loads(always.stdout)
    if document["project"] != "map-base":
        raise AssertionError(document["project"])
    progress = always.stderr.decode("utf-8")
    for phase in ("discovery", "selected", "providers", "joining", "rendering", "complete"):
        if f"] {phase}:" not in progress:
            raise AssertionError(f"missing {phase} phase: {progress}")
    if automatic.stderr:
        raise AssertionError(f"auto progress wrote off-TTY: {automatic.stderr!r}")
    invalid_timeout = subprocess.run(
        [
            sys.executable,
            "-m",
            "archbird",
            "map",
            str(root / "test/fixtures/map_base"),
            "--no-config",
            "--python-provider-timeout",
            "0",
        ],
        cwd=root,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )
    if (
        invalid_timeout.returncode != 2
        or b"--python-provider-timeout must be finite and positive"
        not in invalid_timeout.stderr
    ):
        raise AssertionError(
            f"invalid provider timeout was not rejected: {invalid_timeout!r}"
        )
    cache_parent = root / "build"
    cache_parent.mkdir(exist_ok=True)
    with tempfile.TemporaryDirectory(dir=cache_parent) as raw:
        cache_root = Path(raw) / "cache"
        cached = run(root, "never", cache_dir=cache_root, check=False)
        if cached.stdout != always.stdout or not tuple(
            (cache_root / "maps-v1").rglob("*.json")
        ):
            raise AssertionError("unchecked streaming CLI did not populate Map cache")
        bounded = run(
            root,
            "never",
            cache_dir=Path(raw) / "bounded",
            cache_max_bytes=1,
            check=False,
        )
        if (
            bounded.stdout != always.stdout
            or b"canonical Map exceeded the configured cache budget"
            not in bounded.stderr
        ):
            raise AssertionError("streaming CLI did not report bounded Map cache")
    verify_adaptive_terminal()
    print("Python CLI progress isolation passed")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: test/test_cli_progress.js
#!/usr/bin/env node
"use strict";

const assert = require("node:assert/strict");
const fs = require("node:fs");
const path = require("node:path");
const { spawnSync } = require("node:child_process");

if (process.argv.length !== 5) {
  throw new Error("usage: test_cli_progress.js CLI REPOSITORY ADDON");
}
const cli = path.resolve(process.argv[2]);
const repository = path.resolve(process.argv[3]);
const addon = path.resolve(process.argv[4]);
const fixture = path.join(repository, "test/fixtures/map_base");

function run(mode) {
  const result = spawnSync(process.execPath, [
    cli, "map", fixture, "--config", path.join(fixture, "archbird.json"),
    "--progress", mode, "--no-cache", "--format", "json", "--check",
  ], {
    encoding: "utf8",
    env: {
      ...process.env,
      ARCHBIRD_ENGINE: "native",
      ARCHBIRD_NATIVE_ADDON: addon,
    },
  });
  if (result.status !== 0) {
    throw new Error(`Node progress CLI failed: ${result.stderr}`);
  }
  return result;
}

const always = run("always");
const automatic = run("auto");
assert.equal(always.stdout, automatic.stdout);
assert.equal(JSON.parse(always.stdout).project, "map-base");
for (const phase of ["discovery", "selected", "providers", "joining", "rendering", "complete"]) {
  assert.match(always.stderr, new RegExp(`\\] ${phase}:`));
}
assert.equal(automatic.stderr, "");

const coherenceDirectory = fs.mkdtempSync(
  path.join(repository, "build", "query-coherence-"),
);
try {
  const currentMapPath = path.join(coherenceDirectory, "current-map.json");
  const mismatchedMapPath = path.join(coherenceDirectory, "mismatched-map.json");
  fs.writeFileSync(currentMapPath, always.stdout);
  const mismatched = JSON.parse(always.stdout);
  mismatched.tool.implementation_sha256 = "0".repeat(64);
  fs.writeFileSync(mismatchedMapPath, JSON.stringify(mismatched));
  const runQuery = (mapPath, check) => spawnSync(process.execPath, [
    cli, "query", "--map", mapPath, "--path", "js", "--depth", "0",
    "--format", "json", ...(check ? ["--check"] : []),
  ], {
    encoding: "utf8",
    env: {
      ...process.env,
      ARCHBIRD_ENGINE: "native",
      ARCHBIRD_NATIVE_ADDON: addon,
    },
  });
  const checkedCurrent = runQuery(currentMapPath, true);
  assert.equal(checkedCurrent.status, 0, checkedCurrent.stderr);
  assert.equal(JSON.parse(checkedCurrent.stdout).artifact, "query");
  const brief = spawnSync(process.execPath, [
    cli, "query", "--map", currentMapPath, "--path", "js", "--depth", "0",
    "--view", "changes", "--detail", "compact",
  ], {
    encoding: "utf8",
    env: {
      ...process.env,
      ARCHBIRD_ENGINE: "native",
      ARCHBIRD_NATIVE_ADDON: addon,
    },
  });
  assert.equal(brief.status, 0, brief.stderr);
  assert.match(brief.stdout, /^# Change brief:/);
  assert.match(brief.stdout, /## Evidence limits/);
  const blocked = runQuery(mismatchedMapPath, true);
  assert.equal(blocked.status, 1, blocked.stderr);
  assert.equal(blocked.stdout, "");
  assert.match(blocked.stderr, /saved Map core .* does not match active core/);
  const missingProducerMapPath = path.join(
    coherenceDirectory,
    "missing-producer-map.json",
  );
  const missingProducer = JSON.parse(always.stdout);
  delete missingProducer.tool.implementation_sha256;
  fs.writeFileSync(missingProducerMapPath, JSON.stringify(missingProducer));
  const missingProducerBlocked = runQuery(missingProducerMapPath, true);
  assert.equal(missingProducerBlocked.status, 1, missingProducerBlocked.stderr);
  assert.equal(missingProducerBlocked.stdout, "");
  assert.match(missingProducerBlocked.stderr, /digest is missing or invalid/);
  const crossVersion = runQuery(mismatchedMapPath, false);
  assert.equal(crossVersion.status, 0, crossVersion.stderr);
  const crossVersionDocument = JSON.parse(crossVersion.stdout);
  assert.equal(
    crossVersionDocument.source_tool.implementation_sha256,
    "0".repeat(64),
  );
  assert.notEqual(
    crossVersionDocument.tool.implementation_sha256,
    crossVersionDocument.source_tool.implementation_sha256,
  );
} finally {
  fs.rmSync(coherenceDirectory, { force: true, recursive: true });
}
console.log("Node CLI progress isolation passed");
